### ISSUE TITLE
The abnormal memory usage with zero3 and why can‘t I use zero+offload stratagy?

### ISSUE THREAD

**Issue by dyyoungg** [microsoft/DeepSpeed#4380]
I build deepspeed from the source code, and the version is 0.10.4+dcd3ae1
here is the ds_report. 
```
DeepSpeed C++/CUDA extension op report
--------------------------------------------------
NOTE: Ops not installed will be just-in-time (JIT) compiled at
      runtime if needed. Op compatibility means that your system
      meet the required dependencies to JIT install the op.
--------------------------------------------------
JIT compiled ops requires ninja
ninja .................. [OKAY]
--------------------------------------------------
op name ................ installed .. compatible
--------------------------------------------------
async_io ............... [NO] ....... [OKAY]
fused_adam ............. [NO] ....... [OKAY]
cpu_adam ............... [YES] ...... [OKAY]
cpu_adagrad ............ [NO] ....... [OKAY]
 [WARNING]  Please specify the CUTLASS repo directory as environment variable $CUTLASS_PATH
evoformer_attn ......... [NO] ....... [NO]
fused_lamb ............. [NO] ....... [OKAY]
quantizer .............. [NO] ....... [OKAY]
random_ltd ............. [NO] ....... [OKAY]
 [WARNING]  please install triton==1.0.0 if you want to use sparse attention
sparse_attn ............ [NO] ....... [NO]
spatial_inference ...... [NO] ....... [OKAY]
transformer ............ [NO] ....... [OKAY]
stochastic_transformer . [NO] ....... [OKAY]
transformer_inference .. [NO] ....... [OKAY]
--------------------------------------------------
No CUDA runtime is found, using CUDA_HOME='/mnt/cache/share/cuda-11.7'
DeepSpeed general environment info:
torch install path ............... ['/mnt/cache/env/miniconda3.10/lib/python3.10/site-packages/torch']
torch version .................... 1.13.1+cu117
deepspeed install path ........... ['/mnt/cache/xxx/.local.pt1.13v3/lib/python3.10/site-packages/deepspeed']
deepspeed info ................... 0.10.4+dcd3ae1, dcd3ae1, master
torch cuda version ............... 11.7
torch hip version ................ None
nvcc version ..................... 11.7
deepspeed wheel compiled w. ...... torch 1.13, cuda 11.7
shared memory (/dev/shm) size .... 251.81 GB
```
The main code is as follows
```
parser = HfArgumentParser((CustomizedArguments, TrainingArguments))
args, training_args = parser.parse_json_file(json_file=train_args_file)
model = AutoModelForCausalLM.from_pretrained(
        args.model_name_or_path,
        torch_dtype=torch.float16,
        trust_remote_code=True
    )
if hasattr(model, "enable_input_require_grads"):
      model.enable_input_require_grads()
else:
      def make_inputs_require_grad(module, input, output):
            output.requires_grad_(True)
      model.get_input_embeddings().register_forward_hook(make_inputs_require_grad)

model.gradient_checkpointing_enable()
model.config.use_cache = False
tokenizer = AutoTokenizer.from_pretrained(args.model_name_or_path,trust_remote_code=True)
loss_func = TargetLMLoss(ignore_index=-100)
train_dataset = SFTDataset(args.train_file, tokenizer, args.max_seq_length)
data_collator = SFTDataCollator(tokenizer, args.max_seq_length)
trainer = Trainer(
        model=model,
        args=training_args,
        train_dataset=train_dataset,
        # tokenizer=tokenizer,
        data_collator=data_collator,
        compute_loss=loss_func
    )
trainer.train()
```
Although I can run the code using Zero3, I've noticed that the GPU memory usage is not behaving normally. When fine-tuning the 7B model Qwen on 8 A100 GPUs with 40GB each, the GPU memory usage is as follows. I run my code on slurm  with the instruction
```
srun -p xx --gres=gpu:8 deepspeed --num_gpus=8 --master_port=9901 train.py  --train_args_file xxx.json
```
Here's train_args_file:
```
{
    "output_dir": "/xxxx/results",
    "model_name_or_path": "base_model/Qwen-7b",
    "train_file": "/mnt/cache/yangdeyu/datasets/book/ThreeBody1.txt",
    "task_type":"bookpretrain",
    "deepspeed": "/train_args/deepspeed/deepspeed_zero3.json",
    "num_train_epochs": 50,
    "per_device_train_batch_size": 1,
    "gradient_accumulation_steps": 1,
    "learning_rate": 4e-5,
    "max_seq_length": 1024,
    "logging_steps": 100,
    "save_steps": 500,
    "save_total_limit": 1,
    "lr_scheduler_type": "constant_with_warmup",
    "warmup_steps": 500,
   
    "gradient_checkpointing": true,
    "disable_tqdm": false,
    "seed": 42,
    "bf16": true,
    "report_to": "tensorboard",
    "dataloader_num_workers": 5,
    "save_strategy": "steps",
    "weight_decay": 0,
    "max_grad_norm": 1.0,
    "remove_unused_columns": false
}
```
Here's zero3 configs:
```
{
  "gradient_accumulation_steps": "auto",
  "train_micro_batch_size_per_gpu": "auto",
  "train_batch_size": "auto",
  "gradient_clipping": "auto",
  "zero_allow_untested_optimizer": true,
  "zero_force_ds_cpu_optimizer": false,
  "low_cpu_mem_usage": false,
  "wall_clock_breakdown": false,
  "zero_optimization": {
    "stage": 3,
    "contiguous_gradients": true,
    "allgather_bucket_size": 2e8,
    "reduce_bucket_size": 2e8,
    "overlap_comm": true,
    "reduce_scatter": true,
    "stage3_prefetch_bucket_size": "auto",
    "stage3_param_persistence_threshold": "auto",
    "stage3_gather_16bit_weights_on_model_save": true
  },
  
  "optimizer": {
    "type": "AdamW",
    "params": {
        "lr": "auto",
        "betas": "auto",
        "eps": "auto",
        "weight_decay": "auto"
      }
    },
  
  "scheduler": {
      "type": "WarmupLR",
      "params": {
          "warmup_min_lr": "auto",
          "warmup_max_lr": "auto",
          "warmup_num_steps": "auto"
      }
    },

  "bf16": {
    "enabled": true
  }
}

```
Here's GPU memory usage. It can be observed that each card is using around 38GB of memory, which I find to be abnormal when training with Zero3 in bfloat16 precision, considering that Zero3 incorporates model and gradient parallelism. Especially the batchsize is 1 and max_seq_length is 1024.
```
0  NVIDIA A100-SXM...  On   | 00000000:26:00.0 Off |                    0 |
| N/A   54C    P0   368W / 400W |  38079MiB / 40960MiB |     93%      Default |
|                               |                      |             Disabled |
+-------------------------------+----------------------+----------------------+
|   1  NVIDIA A100-SXM...  On   | 00000000:2C:00.0 Off |                    0 |
| N/A   60C    P0   377W / 400W |  37579MiB / 40960MiB |     93%      Default |
|                               |                      |             Disabled |
+-------------------------------+----------------------+----------------------+
|   2  NVIDIA A100-SXM...  On   | 00000000:65:00.0 Off |                    0 |
| N/A   61C    P0   291W / 400W |  37569MiB / 40960MiB |     93%  
```

I think that there's something wrong with deepspeed but I haven't find the causes yet. When I use zero3+offload，the subprocess will be killed after parameter offload.
```
2023-09-21 11:40:57.390 | INFO     | __main__:main:245 - *** starting training ***
Parameter Offload: Total persistent parameters: 659456 in 97 params
[2023-09-21 11:42:33,233] [INFO] [launch.py:315:sigkill_handler] Killing subprocess 117432
.....
```
It seems that the process get stuck in the following code

```
File "xxxx/lib/python3.10/site-packages/deepspeed/launcher/runner.py", line 586, in main
    result.wait()
File "/env/miniconda3.10/lib/python3.10/subprocess.py", line 1209, in wait
    return self._wait(timeout=timeout)
File "/env/miniconda3.10/lib/python3.10/subprocess.py", line 1959, in _wait
    (pid, sts) = self._try_wait(0)
File "/env/miniconda3.10/lib/python3.10/subprocess.py", line 1917, in _try_wait
    (pid, sts) = os.waitpid(self.pid, wait_flags)
```

**Comment by jomayeri** [comment]
@dyyoungg The process is being killed during offload because it is consuming to much CPU memory. The GPU memory usage also appears reasonable for such a large model and longer sequence length.
